### PR TITLE
implement an encode method

### PR DESCRIPTION
**Pull request by omarandlorraine**
We could use an `encode` method on an `Instruction` to yield the instruction's encoding.

I've essentially copy-pasted the function I'm using into the yaxpeax-6502 codebase. But it doesn't build because `macro vec not found in this scope`? Why is this happening? Maybe it's something to do with no-std. I've never seen this error before. If I can't use `Vec<u8>` for this then I'll probably do a custom type that `impl`s `Iterator<u8>`. But this decision needs to be considered carefully and with regard for what other yaxpeax backends might need.

Still other things to consider:

- [ ] testing decoder and encoder against each other
- [ ] Some kind of trait that's usable on other yaxpeax backends as well; this probably needs involvement from the `yaxpeax-arch` project

### CODE REVIEW

**Comment by cr1901** [comment]
>Maybe it's something to do with no-std. I've never seen this error before. If I can't use Vec<u8> for this then I'll probably do a custom type that impls Iterator<u8>.

Yes, this is due to `no_std` without alloc. Please return a type that impls `Iterator<u8>`; if users want a `Vec`, they can `collect` into one in their crates that uses `std` or `alloc`.

>But this decision needs to be considered carefully and with regard for what other yaxpeax backends might need.

The backends are nominally `no_std`, and I believe the only reason a `std` feature exists is because the `Error` trait isn't in `core` yet. Additionally, Idk if the maintainer wants to add an encoding trait. I'm not personally too concerned with doing a breaking change if the first attempt at a `encode` method doesn't look like a hypothetical `Encoder` trait down the road.

**Comment by omarandlorraine** [comment]
Okay, so `encode` now returns a new type that impls `Iterator<u8>`

**Comment by cr1901** [comment]
@omarandlorraine Two things, and then this is fine/I'll accept.

1. Can we rename `ByteByByte` => `Bytes` (_this is a bikeshed_)?
2. Can we return an `EncodingError` `enum` on `match` failure instead of `panic`king?

**Review by cr1901 — APPROVED** [review]

**Comment by cr1901** [comment]
LGTM, tyvm for the PR :D!

**Comment by omarandlorraine** [comment]
I've just made the requested changes, but I don't see the two commits turn up here. Maybe because the pull request has already been closed?

**Comment by cr1901** [comment]
I'm not sure what could be happening on your end; all your changes incl. the two commits were merged into the main branch just fine on my end.

**Comment by omarandlorraine** [comment]
> So on the encoder branch of omarandlorraine/yaxpeax-6502 we've got two commits that have not made it into your main branch: 
> 
> - `8fcc2ee` return Err(EncodeError) instead of panicking
> - `67175c7` rename decode to encode
> 
> they are the responses to your code review
> 
> Should I open another PR maybe?

I am aware that these commits were pushed after the PR was merged

**Comment by cr1901** [comment]
>Should I open another PR maybe?

Open another PR. That was my fault; I thought you were done.

**Pull request merged**
